### PR TITLE
Replace unused BIP 9 logic with draft BIP 8

### PR DESCRIPTION
**Pull request by luke-jr**
Since all BIP 9 deployments are buried, we can simply replace it with BIP 8 cleanly.

### CODE REVIEW

**Comment by luke-jr** [comment]
Rebased and updated with some of @aj's changes to BIP 8.

This is probably still not the final form of BIP 8 (see https://github.com/bitcoin/bips/pull/950), but merging this sooner means we have fewer changes to backport together with the activation parameters later on.

(It also means the RPC interfaces are updated now instead of later)

File: src/chainparams.cpp
Comment (ajtowns) [inline src/chainparams.cpp:396]: `startheight = 144` would be more accurate, it can't actually start at height 0.

(This caused a weird bug in `validation_blocks_tests` when I tried implementing bitcoin/bips#1023 and it was able to go to STARTED in the first block; didn't track down exactly what the issue was though)

**Review by ajtowns — COMMENTED** [review]
This has some bugs, see below.

I've done a branch with my suggestions at https://github.com/ajtowns/bitcoin/tree/202010-bip8-suggestions -- it does the following:
* makes the output of the versionbits unit test more helpful when it finds bugs
* updates the unit tests to check the "never active" case (which catches a bug in this PR)
* adapts your commits for the changes above, since there were merge conflicts
* has some explicit "fixups" for your commits fixing bugs or updating the changes above for your patches
* has a few refactors for versionbits and bip8 support that I think makes things a bit easier to follow
* has draft implementations for the open PRs against BIP8

**Comment by michaelfolkson** [comment]
Concept ACK. 

Since revised BIP 8 is a strict improvement on BIP 9 it doesn't hurt to replace BIP 9 logic with revised BIP 8 logic. AJ has [highlighted](https://github.com/bitcoin/bitcoin/pull/19573#pullrequestreview-511393873) some bugs, improvements that haven't been addressed yet. I also don't know if you are waiting for revised BIP 8 to be finalized at this point before a rebased version of this is ready to be merged.

**Comment by luke-jr** [comment]
Rebased, merged in @ajtowns's fixes and other changes, and got CI passing.

File: src/rpc/blockchain.cpp
Comment (ajtowns) [inline src/rpc/blockchain.cpp:1245]: Test should be `thresholdState != ThresholdState::LOCKED_IN` -- now that https://github.com/bitcoin/bips/pull/1021/ is merged, threshold is still relevant during MUST_SIGNAL phase.

**Review by ajtowns — COMMENTED** [review]
Sample functional test at https://github.com/ajtowns/bitcoin/commits/202102-bip8-functest -- it only picked up a minor issue with getblockchaininfo

File: src/chainparams.cpp
Comment (benthecarman) [inline src/chainparams.cpp:493]: this log should probably include the `timeoutheight`

File: src/consensus/params.h
Comment (benthecarman) [inline src/consensus/params.h:31]: >  Should be a retarget block

In BIP 8, it defines these as a *must*


> startheight and timeoutheight must be an exact multiple of 2016 (ie, at a retarget boundary), and timeoutheight must be at least 4096 blocks (2 retarget intervals) after startheight.

File: src/rpc/mining.cpp
Comment (benthecarman) [inline src/rpc/mining.cpp:830]: > sed

set

File: src/test/versionbits_tests.cpp
Comment (benthecarman) [inline src/test/versionbits_tests.cpp:23]: Couldn't this use a `switch` statement?

File: src/chainparams.cpp
Comment (luke-jr) [reply src/chainparams.cpp:493]: You mean `lockinontimeout`, I assume?

File: src/test/versionbits_tests.cpp
Comment (luke-jr) [reply src/test/versionbits_tests.cpp:23]: This PR is intended to merge cleanly to 0.21, which only requires C++11. (`switch` inside `constexpr` functions requires C++14)

**Comment by luke-jr** [comment]
Addressed review comments & added @ajtowns 's new test.

Intentionally left threshold at 95% for now - that is best left for a future PR to change (when defining a softfork to use BIP8)

File: src/chainparams.cpp
Comment (benthecarman) [reply src/chainparams.cpp]: Yes, sorry

**Review by benthecarman — APPROVED** [review]
ACK bd8517135fc839c3332fea4d9c8373b94c8c9de8

**Comment by TheBlueMatt** [comment]
I'm a little dubious that forced-signaling logic shouldn't come with code to handle the network forking (ie more aggressive peer-searching via protocol magic or quicker new connections when the chain has stalled). It seems like forced-signaling (which happily forks itself off the network in many of the ways people are talking about using the code today) probably isn't something that should be merged unless the code for it is complete in the sense that it handles likely outcomes gracefully.

**Review by darosior — COMMENTED** [review]
Concept ACK -- considerations on the parameters aside, BIP8 is a strict improvement over BIP9. Will review soon.

File: src/rpc/blockchain.cpp
Comment (prayank23) [inline src/rpc/blockchain.cpp:1295]: nit: _"the minimum height of a block at which the bit gains its meaning"_ can be changed to one of the below

_"the minimum height of a block at which the 'bit' gains its meaning"_
_"the minimum height of a block at which the 'bit' becomes relevant"_
_"the minimum height of a block at which the 'bit' used for signaling gains its meaning"_

File: src/validation.cpp
Comment (lucasmoten) [inline src/validation.cpp]: Referencing BIP9Stats?

Comment (lucasmoten) [reply src/validation.cpp]: as a local it should be harmless

**Comment by lucasmoten** [comment]
utACK

**Comment by prayank23** [comment]
Concept ACK. Need to do more research and test everything soon.

**Comment by michaelfolkson** [comment]
Based on today's code review [session](http://gnusha.org/taproot-activation/2021-02-23.log) on IRC I think I can ACK the non-test commits.

I've got an assertion error on the feature_bip8.py functional test when run locally which I will look into tomorrow. 

`AssertionError: [node 3] Expected messages "['bad-vbit-unset-testdummy']" does not partially match log:`

> I'm a little dubious that forced-signaling logic shouldn't come with code to handle the network forking (ie more aggressive peer-searching via protocol magic or quicker new connections when the chain has stalled).

I'm not sure there is much we can do in the (worst case) scenario of a chain split due to a proportion of the network running lot=true and a proportion of the network running lot=false during the MUST_SIGNAL phase. I can see there is a desire to do **something** as it is a worry but I wonder if these suggested P2P changes would potentially do as much harm as good. For example a lot=x node would want to find out if they were on the wrong side of network consensus (lot=y) rather than just pairing with lot=x nodes.

However, I think that even if revised BIP 8 didn't have the lot=true option there would still be demand for a UASF solution in certain scenarios and hence that is why I originally ACKed the changes to BIP 8 to include a lot=true option. Better to have it within a flexible BIP (with parameters) rather than a UASF effort just grabbing a random BIP off the shelf.

**Comment by Sjors** [comment]
This or a followup might be a good opportunity to test p2p behavior against older versions, see `feature_backwards_compatibility.py` for inspiration. You can also use this framework to test the interaction between `lockinontimeout` `true/false` nodes. Though I agree with @TheBlueMatt that this PR would be simpler is we only support `false` for now.

**Comment by benthecarman** [comment]
ACK `977280c5f04c70c2610ef5fcfde57e5256171f65`

**Comment by AnthonyRonning** [comment]
ACK 977280c5f04c70c2610ef5fcfde57e5256171f65 based on the discussions in yesterday's IRC chat and some additional code review & running the tests locally.

**Comment by michaelfolkson** [comment]
In light of a [project](https://github.com/BitcoinActivation/bitcoin) set up to implement a default of lot=true I think Core should implement a default of lot=false with no user configuration option. I also think there should be signposting (comments etc) of where lot=true logic would go in the Core codebase (and referring to the lot parameter in BIP 8) but I don't think it makes sense to have any lot=true logic in the Core codebase that will remain unused for Taproot activation. 

The alternative is Core doesn't release any activation code and entirely leaves it to non-Core releases. Personally I don't think this is optimal. I think some users will want to run a Core release that can activate Taproot with miner signaling without being forced to run a non-Core release to get Taproot activated. A significant proportion (unclear whether majority) of the community seems perfectly happy with lot=false as a default and have no desire to change it.

In the worst case scenario of miners failing to activate within say the first 6 months we will need to take the chain split risk at the end the year (2 weeks before timeoutheight) more seriously (and consider providing tooling and communications to Core users well in advance). I think the highly likely scenario in Taproot's case is miners activate in the first 6 months and the lot parameter ends up being entirely irrelevant.

**Comment by luke-jr** [comment]
No, the alternative is Core release LOT=true and stay consensus-compatible instead of intentionally harming Core's users.

**Comment by JeremyRubin** [comment]
I think that it is a really bad idea to release LOT=false without the ability to trivially set LOT=true. Should LOT=true become widespread, doing things with less time creates risk of a 'forced error' on activation. It's entirely possible that miners will have proprietary mining software that they would need to rebase onto a recompiled client, and that might be a bottleneck. I suppose that LOT=false -> LOT=true patches would be small, but it seems to me much safer to release a client that *could* maintain consensus with a network that is LOT=true, rather than having to take the hit on the software upgade cycles necessitated by a re-release.


My suggestion (which perhaps belongs on the mailing list!) is that if in aggregate the sentiment of core contributors is that it is a bad idea to release with LOT=true available via configuration, we could make it such that LOT=true being set requires LOTHASH = block hash at height startheight + 6 months of blocks. This way clients provably cannot be configured with LOT=true unless they restart their node with the config parameter determined in the future. This reduces the client "supply chain risk" engendered by making users recompile / switch binaries.

**Comment by luke-jr** [comment]
Also, as a reminder, this PR is not about setting activation parameters, only merging BIP 8 code.

**Comment by michaelfolkson** [comment]
> Also, as a reminder, this PR is not about setting activation parameters, only merging BIP 8 code.

That's true but we do have Matt and Sjors both commenting that this PR should perhaps only include lot=false code hence the reason for my comment. I wouldn't seek to add noise for no reason.

**Comment by luke-jr** [comment]
There is no reason to include LOT=False code and not LOT=True code also. In any case, this PR has sufficient ACKs and is ready to merge today.

**Comment by TheBlueMatt** [comment]
Let's keep this discussion to code-level - broad debates about what Core should or shouldn't include and philosophical debates probably belong on the ML. My comment was that it doesn't make sense to include highly feature-incomplete code for LOT=true (which is the case in this PR). It may make sense to just merge a LOT=false PR, or to delay this until there is code to properly handle the network splitting case implied by a LOT=true option.

**Comment by JeremyRubin** [comment]
@luke-jr / @TheBlueMatt  do you have any insight on how urgent improving peer selection is in any case? Are the usual mechanisms for dealing with nodes which mine an invalid tip (from the pov of a lot=true) or nodes which send stale headers (pov of lot=false) insufficient?

It's not immediately obvious to me that the network as-is wouldn't be tolerably resilient to this sort of issue.

**Comment by luke-jr** [comment]
LOT=True does not differ from any other softfork in respect to handling network divergence. Perhaps the latter can be improved, but it is out of scope to this PR.

**Comment by benthecarman** [comment]
> LOT=True does not differ from any other softfork in respect to handling network divergence. Perhaps the latter can be improved, but it is out of scope to this PR.

Agree with luke-jr here, if you are running a lot=true client then you deem the old chain invalid

**Comment by TheBlueMatt** [comment]
An option for LOT=true and an expectation that nodes on the network will be running different consensus rules implies something very, very different from any previous fork. Indeed, during the Segwit2x/etc debacle some emergency changes were rushed in to hopefully make the P2P network slightly more robust against such things, but they're fare from an ideal scenario, and if we're going into such a case with eyes open, there's no question they should be improved significantly.

**Comment by luke-jr** [comment]
This PR does not add such a (user-accessible) option, nor does it set any expectation for divergent rules.

**Comment by TheBlueMatt** [comment]
I wasn't leaning on the PR's current code for my expectation of divergent rules :).

**Comment by michaelfolkson** [comment]
ACK 977280c5f04c70c2610ef5fcfde57e5256171f65

I'm getting a little concerned that some individuals will continuously shift the goalposts in search of a philosopher's stone for an activation mechanism that has 100 percent consensus or a philosopher's stone to magic away the (unlikely but still possible) risks of a chain split after a year of miners failing to activate. 

Anyone who doesn't like this PR is free to open an alternative PR.

> Are the usual mechanisms for dealing with nodes which mine an invalid tip (from the pov of a lot=true) or nodes which send stale headers (pov of lot=false) insufficient? It's not immediately obvious to me that the network as-is wouldn't be tolerably resilient to this sort of issue. 

This question needs to be answered with more specifics than just "they should be improved significantly." And then those suggestions should ideally be implemented in an independent PR so there can be discussion on whether they are indeed P2P improvements in a chain split scenario or just additional complexity. That would certainly be more productive than creating a logjam on any activation PRs that are opened.

File: src/validation.cpp
Comment (ajtowns) [inline src/validation.cpp:3566]: I don't think the `++stats.count` there is correct -- if the final block in the MUST_SIGNAL period does not signal, and the total number of blocks signalling in the period is `threshold-1`, this will cause lockinontimeout=true nodes to transition from MUST_SIGNAL to LOCKED_IN (and later to ACTIVE) while lockinontimeout=false nodes transition from STARTED to FAILED.

Setting `N=107` instead of a random value and removing the `random.shuffle(bits)` in the test case should duplicate the failure.

**Comment by TheBlueMatt** [comment]
> This question needs to be answered with more specifics than just "they should be improved significantly."

I did, read above :)

**Comment by michaelfolkson** [comment]
> More broadly, it seems the discussion on the mailing list and elsewhere is moving beyond BIP 8

I guess if you don't engage in the discussion for 6 months as the alternatives are outlined, discussed and narrowed down it might seem that way. From what I see a UASF LOT=true release effort is beginning. But you are free to start the 6 month long activation discussion from scratch in search for your philosopher's stone if you feel that is a good use of your time.

**Comment by Sjors** [comment]
I'm not opposed to `LOT=True` _code_, but I'm (currently) not interesting in developing, reviewing or testing it. I might be, if there was a PR specifically dedicated to adding such functionality. But this PR holds back `LOT=False` support, by making it a package deal.

I'm also opposed to merging this without said review. The ACKs so far are not from people with sufficient expertise to judge soft-fork activation code (I don't consider myself qualified either).

I'm happy to review `LOT=False` code. I may also extract a few bits from this PR into separate PR's so the review burden decreases for everyone.

Jeremy Rubin wrote:
> I think that it is a really bad idea to release LOT=false without the ability to trivially set LOT=true.

This is certainly not ideal, but perhaps that suggest a problem with the LOT parameter in BIP 8 introducing too much complexity.

**Comment by michaelfolkson** [comment]
> I'm also opposed to merging this without said review. The ACKs so far are not from people with sufficient expertise to judge soft-fork activation code (I don't consider myself qualified either).

Totally agree. This (or an alternative PR) desperately needs long term contributors reviewing the code. I suspect the reason why they are delaying reviewing the code (or opening an alternative PR) is an unwillingness to get involved in the high level discussion (which has been done to death imo). With the effective NACKs by Core contributors I have seen (we can list them if this is helpful) on releasing a default LOT=true I can't see any route to Core releasing a default of LOT=true. It is a default of LOT=false or releasing nothing (or another 6-8 months of searching for the philosopher's stone of course).

> I'm happy to review LOT=False code. I may also extract a few bits from this PR into separate PR's so the review burden decreases for everyone.

This would be great.

File: src/test/versionbits_tests.cpp
Comment (Sjors) [inline src/test/versionbits_tests.cpp]: bae9a452191a7a83478f7d508a54f4a04d385505  and bd715ff894344408df2345d9bd75a08bd7c9c6f6 : why the `+ 1`? The log will show `0` for the genesis block, but 2 the block at height 1.

**Comment by Sjors** [comment]
Some additional notes:
@TheBlueMatt [wrote](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-February/018450.html) on the mailinglist:

> Bitcoin Core does not have infrastructure to handle switching consensus rules with the same datadir - after running with 
uasf=true for some time, valid blocks will be marked as invalid, and additional development would need to occur to 
enable switching back to uasf=false. This is complex, critical code to get right, and the review and testing cycles 
needed seem to be not worth it.

This scenario can be tested by stopping the `-vbparams=testdummy:@{start}:@{stop}:1` node in `feature_bip8.py` and then restarting it with `-vbparams=testdummy:@{start}:@{stop}:1` (and vice versa for someone opting in later. 

A more practical solution might be to require a user who changes their mind to resync the chain. But you'd have to store the choice somewhere.

**Comment by TheBlueMatt** [comment]
> I guess if you don't engage in the discussion for 6 months as the alternatives are outlined, discussed and narrowed down it might seem that way.

Lets keep the debates to the ML. I'm not gonna spam this PR with personal attacks.

> This scenario can be tested by stopping the -vbparams=testdummy:@{start}:@{stop}:1 node in feature_bip8.py and then restarting it with -vbparams=testdummy:@{start}:@{stop}:0 (and vice versa for someone opting in later.

Right, I believe AJ mentioned in that thread that it may (mostly) work already (though that certainly needs careful thought!), however there's not a good way to test the P2P aspects of a network split, and that's certainly also important to handle for a BIP8-based fork with strong community views.

**Comment by luke-jr** [comment]
Fixed the bugs @ajtowns and @Sjors  found

**Comment by sdaftuar** [comment]
I'm a Concept NACK on BIP 8 as drafted.  I think the MUST_SIGNAL period is needlessly risky for maintaining consensus (as I wrote back in 2017 on the mailing list, responding to a similar proposal at the time); if that part of BIP 8 were dropped I'd be fine with an implementation of the rest, though I haven't reviewed the code here.

**Comment by michaelfolkson** [comment]
@sdaftuar: I'm sure you already know this but just to ensure clarity. BIP 8 (LOT = false) does not have the MUST_SIGNAL period and hence an implementation of BIP 8 (LOT = false) does not have the MUST_SIGNAL period.

If you want MUST_SIGNAL or the LOT parameter stripped from the actual BIP you'll need to open a BIP PR. I don't personally see what that would achieve unless you think LOT=true and MUST_SIGNAL should not be used in **any** circumstance whatsoever by either Core or an alternative implementation.

File: src/test/versionbits_tests.cpp
Comment (ajtowns) [reply src/test/versionbits_tests.cpp]: `height` here is the height of the next block -- if `vpblock` is empty, the next (ie first) block should be the genesis block with `nHeight == 0`. Note that when populating `vpblock` in `VersionBitsTester::Mine()` we set `nHeight=vblock.size()` so the first entry will have `nHeight == 0` and act as the genesis block; at which point `1+vpblock.back().nHeight` will evaluate to `1+0` giving the height of the next block to be added. When `vpblock` is empty, it'll evaluate to `0`, still giving the height of the next block to be added.

Comment (ajtowns) [reply src/test/versionbits_tests.cpp]: Note that `GetStateFor(block)` is telling you what the state should be for the *next* block to be mined, not the last block that was mined.

**Comment by Sjors** [comment]
@TheBlueMatt wrote:
> there's not a good way to test the P2P aspects of a network split

We can now test using binaries of earlier releases, so this might be more feasible. But I'd rather see that in a PR that's specifically focussed on LOT. (my other objections aside)

**Comment by michaelfolkson** [comment]
It appears this PR doesn't have much (any?) chance of being merged in its current form (ie whilst it has any LOT=true code). Hence I (and others I have discussed it with) think it is best if this PR is closed.

An approach like Sjors' recent PR https://github.com/bitcoin/bitcoin/pull/21334 stripping out particular commits seem the best way forward.

I'll reiterate my view (and I know Luke is strongly opposed) that if Core ships any activation code it will be BIP 8(1 year, false). 

I also highly doubt (though people are welcome to try by opening a BIP PR) that the LOT parameter will be stripped out of BIP 8 given the work and the consensus has built around it over the past weeks. But a LOT=false implementation in Core does not need any LOT code.

File: src/test/versionbits_tests.cpp
Comment (Sjors) [reply src/test/versionbits_tests.cpp]: I see, I'll add a comment to that effect in my PR.

**Comment by ProofOfKeags** [comment]
@michaelfolkson Are you saying that the LOT=true code can't even exist, or just that it can't be enabled for any SF proposals (specifically Taproot)?

Seems like a bizarre requirement considering it is actually in the BIP, but perhaps theres a reason.

**Comment by michaelfolkson** [comment]
@ProofOfKeags: I think it is just semantics. It appears some people don't want LOT=true code in the Core codebase. Obviously it would be dead code if Core implemented the equivalent of LOT=false so it really doesn't matter. The Core codebase and the BIP (BIP 8) are separate objects. I can't see LOT=true code being merged into Core but I also can't see LOT or MUST_SIGNAL getting taken out of the BIP.

Having said that I didn't think there would be an alternative [proposal](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-March/018583.html) ("Speedy trial") gather momentum at this late stage so please take any/all of my predictions with a large pinch of salt :)

**Comment by achow101** [comment]
If 31eac50c721dd3b0bd2347e76196bf16913e9be9 (from #21055) is backported to 0.21, then this PR rebased on master will also backport to 0.21 cleanly.

File: src/versionbits.cpp
Comment (ajtowns) [inline src/versionbits.cpp:240]: Sorry, I think this change is wrong and it should be the other way around -- `AbstractThresholdConditionChecker::GetStateStatisticsFor` should be expecting `pindexPrev` not `pindex`. I believe that gives the right behaviour for MUST_SIGNAL. I'd suggest just dropping this commit.

**Comment by jtimon** [comment]
If I'm understanding the code correctly at a glance, this also gets rid of the LOT=false option and instead provides an option to never activate the change regardless of signaling, which in my opinion makes more sense as it gives users a more meaningful option than "do you want to allow miners to have veto powers over the change and potentially fork off from other users who like you also want this change to be applied?".
Giving users the option of giving miners veto power doesn't empower users at all. if you're super worried about "perception", then give users the option to be against the change directly.

Concept ACK.

But I would actually go further and for the users who select that option, not only they should not activate the change, but also consider the block that activates the change invalid, so that they can cleanly fork off to their own chain without the change they oppose.

**Comment by luke-jr** [comment]
This is being split up and merged in pieces (some modified), so closing it.

**Comment by luke-jr** [comment]
Rebased (CI failure appears unrelated)

**Comment by achow101** [comment]
Note that we may still want to use this in the future. Merging #21377 now does not preclude the use of BIP 8 in a future deployment.

**Comment by MarcoFalke** [comment]
Ok, let me know when to reopen, but I don't think this needs to stay in "High-priority for review"

**Comment by Sjors** [comment]
Agreed, it's probably a good idea to keep working on BIP 8 between now and the next soft fork attempt (especially the LOT=false part). That seems better than to develop these mechanisms only just in time.

But in order to keep backporting easy, I suggest we don't merge a BIP 8 implementation until Speedy Trial expires or locks in.

**Comment by luke-jr** [comment]
Thinking about it, it doesn't make sense to continue this PR here. This has LOT=True support and such, and what is needed right now is simply heights.

When/if the folks impeding progress of BIP 8 change their mind, I can do that. But not sure there's any point rebasing until then.

**Pull request closed without merging**
